### PR TITLE
Add imx3112 mux driver

### PR DESCRIPTION
**Pull request by ramiroderojas**
Change adds driver support for I3C 2:1 mux splitter Renesas IMX3112: https://www.renesas.com/us/en/products/memory-logic/memory-interface-products/memory-multiplexers/imx3112-i3c-basic-12-bus-multiplexer

The first commit adds support for buses with I3C devices ONLY.
Second commit adds support for mixed and I2C-only buses.

Signed-off-by: Ramiro de Rojas Perez ramiroderojas@google.com

### CODE REVIEW

**Comment by ramiroderojas** [comment]
@mfischer Can you take a look :)

File: drivers/i3c/i3c_renesas_imx3112.c
Comment (XenuIsWatching) [inline drivers/i3c/i3c_renesas_imx3112.c:422]: I'm afraid this could have consequences with initialization of other devices on the bus as this is a broadcast message where it cause ALL devices on the bus (if SETAASA supported) to load their static address as their dynamic address regardless of what is intended from the device tree. I don't have the IMX3112 data sheet in front of me... but are you sure this is the only way for this to "switch" to I3C mode. That CCC is just one of the ways of assigning a DA. The controller of the bus this is attached to should be the one already assigning a DA to this mux which should initialized before this which should already be doing as apart of it's init func.
This really does not seem right.

**Review by XenuIsWatching — CHANGES_REQUESTED** [review]
You may want also do what https://github.com/zephyrproject-rtos/zephyr/pull/59163 is doing where it makes sure the 2 priority configs are set right.

File: drivers/i3c/i3c_renesas_imx3112_regs.h
Comment (mfischer) [inline drivers/i3c/i3c_renesas_imx3112_regs.h:5]: Leftover? :)

File: drivers/i3c/i3c_renesas_imx3112.c
Comment (mfischer) [inline drivers/i3c/i3c_renesas_imx3112.c]: I don't think you need the _OR ones for booleans. They're either present and true, or absent and false.

File: drivers/i3c/i3c_renesas_imx3112.c
Comment (mfischer) [inline drivers/i3c/i3c_renesas_imx3112.c:577]: Would DT_INST_FOREACH() an option here? I'm not sure, do 1:4 and 1:8 versions of this exist that might justify a more generic driver?

File: drivers/i3c/i3c_renesas_imx3112.c
Comment (mfischer) [inline drivers/i3c/i3c_renesas_imx3112.c]: Are there higher channel count devices of the same series? If so maybe consider a
i3c_renesas_imx311x or similar name.

File: drivers/i3c/i3c_renesas_imx3112.c
Comment (mfischer) [inline drivers/i3c/i3c_renesas_imx3112.c:95]: Does this have to be a struct? Is this getting re-used somewhere?

File: drivers/i3c/i3c_renesas_imx3112.c
Comment (ramiroderojas) [reply drivers/i3c/i3c_renesas_imx3112.c:422]: Thanks Ryan for taking a look :)

The reason why I don't set dynamic addresses for each of the child devices in the driver code is that all the child devices we use in the mux just use their SA's as DA's, so we never had to set DA's individually (i.e use ENTDAA). I'll investigate how to do other methods of address assignment and see if i can adjust the driver code. 

 I agree that setting the device to I3C should be left to the I3C controller, so we can get rid of this of this SETAASA.

File: drivers/i3c/i3c_renesas_imx3112.c
Comment (ramiroderojas) [reply drivers/i3c/i3c_renesas_imx3112.c:95]: Which struct do you refer to, mux_data or mux_state?

File: drivers/i3c/i3c_renesas_imx3112.c
Comment (ramiroderojas) [reply drivers/i3c/i3c_renesas_imx3112.c:577]: There are no 1:4/1:8 versions of this mux. There are 2:4/2:8 intelligent switches for I3C but those have a completely different API, so it wouldn't really make sense to try to generalize imho

File: drivers/i3c/i3c_renesas_imx3112.c
Comment (ramiroderojas) [reply drivers/i3c/i3c_renesas_imx3112.c]: There aren't

File: drivers/i3c/i3c_renesas_imx3112.c
Comment (XenuIsWatching) [reply drivers/i3c/i3c_renesas_imx3112.c:422]: Yeah... it should, but it appears this may run into a limitation of the MACRO that gathers up all the i3c devices on the bus. I ran in to this issue with the tca9546. Which I wrote about there... https://github.com/zephyrproject-rtos/zephyr/issues/51828

I still haven't figured out the macrobatics yet to solve it :(

File: drivers/i3c/i3c_renesas_imx3112.c
Comment (ramiroderojas) [reply drivers/i3c/i3c_renesas_imx3112.c:95]: If you're talking about the mutex state, i'm just separating the lock functionality from everything else for simplicity.

File: drivers/i3c/Kconfig.renesas
Comment (dcpleung) [inline drivers/i3c/Kconfig.renesas:5]: Both `*_PRIORITY` summary need to be clarified since they are obviously different kconfigs.

File: drivers/i3c/i3c_renesas_imx3112.c
Comment (dcpleung) [inline drivers/i3c/i3c_renesas_imx3112.c:4]: 2023?

File: drivers/i3c/i3c_renesas_imx3112.c
Comment (ramiroderojas) [reply drivers/i3c/i3c_renesas_imx3112.c]: Ah yeah, thank you for adding the attach/detach api https://github.com/zephyrproject-rtos/zephyr/pull/52035. Its unfortunate we haven't found a macrobatic expert to get us a macrobatic i2c/i3c tree yet :/.

As a workaround for now, we have been using the following design pattern:

```
&i3c_controller {
      lps22hh0_mock: lps22hh_mock@10 {
              compatible = "vnd,i3c-device";
              reg = <0x10 0xabcd 0x12345010>;
      };

      /* The imx3112 node must be a child of an i3c controller */
      mux: imx3112@77 {
          compatible = "renesas,imx3112-i3c";
          reg = <0x77 0xabcd 0x02345678>;
          status = "okay";
          #address-cells = <1>;
          #size-cells = <0>;
          mux_i3c@0 {
              compatible: "renesas,imx3112-channel"
              reg = <0>;
              #address-cells = <3>;
              #size-cells = <0>;
              lps22hh0: lps22hh@10 {
                      compatible = "st,lps22hh";
                      reg = <0x10 0xabcd 0x12345010>;
              };
          };
          mux_i3c@1 {
            compatible: "renesas,imx3112-channel"
              reg = <1>;
              #address-cells = <3>;
              #size-cells = <0>;
              lps22hh1: lps22hh@10 {
                      compatible = "st,lps22hh";
                      reg = <0x10 0xabcd 0x12345010>;
              };
          };
      };
};

```
Where the vnd,i3c-device can just be a stub driver.

Obviously this works best if we have the same addresses in each channel, however.

**Review by dcpleung — COMMENTED** [review]
Could you put this one under `drivers/i3c/mux`? This is not exactly a driver for a I3C host controller.

File: drivers/i3c/i3c_renesas_imx3112.c
Comment (ramiroderojas) [reply drivers/i3c/i3c_renesas_imx3112.c:4]: I made the initial commits in 2022, i'm just upstreaming it this year.

File: drivers/i3c/Kconfig.renesas
Comment (ramiroderojas) [reply drivers/i3c/Kconfig.renesas:5]: Good point, let me do that

**Comment by kartben** [comment]
@ramiroderojas will you be able to address the pending comments? Thanks again for the contribution!

File: drivers/i3c/i3c_renesas_imx3112.c
Comment (ramiroderojas) [reply drivers/i3c/i3c_renesas_imx3112.c]: I will resolve this if that's okay.

File: drivers/i3c/Kconfig.renesas
Comment (ramiroderojas) [reply drivers/i3c/Kconfig.renesas]: They are different Kconfigs, but one is useless without the other. 
I hope you will find my new comments explain the situation much better. Thank you!

File: drivers/i3c/i3c_renesas_imx3112.c
Comment (ramiroderojas) [reply drivers/i3c/i3c_renesas_imx3112.c]: Do you think we can resolve this? I am unsure if there's a better way to switch to I3C for this mux's children, since the mux does not support ENTDAA.

Comment (ramiroderojas) [reply drivers/i3c/i3c_renesas_imx3112.c]: @XenuIsWatching I was hoping you could take a final look, thanks :)

Comment (XenuIsWatching) [reply drivers/i3c/i3c_renesas_imx3112.c]: sorry, I don't have the data sheet for this part so I can't really dig into what it supports... (it sounds odd that something can not support ENTDAA it is required for all i3c devices)
I really don't know what mean by "switch to I3C".... do you just mean assign it's DA?

Ideally, how I'd like to see this resolve is for additional Python code to be created to allow for additional macros that would allow the I3C macros to gather the devices in the mux... but then there is the issue of how the address table would handle duplicate SA

I don't believe sending SETAASA is right all here... That is an optional CCC command that devices can support, so it working with all devices can't be guaranteed here.

Comment (mbolivar-ampere) [reply drivers/i3c/i3c_renesas_imx3112.c]: > Maybe [...] @mbolivar-ampere (so many handles here...) has an idea about the python code https://github.com/zephyrproject-rtos/zephyr/issues/61928

This is the correct current account. Sorry for the churn.

I think we talked about this in person. There's no issue conceptually with adding a `DT_FOREACH_DESCENDANT(node_id, fn)` macro that expands to calling `fn` on each descendant of `node_id` (although you might want to have some extra macros to control visitation order -- depth first, breadth first, etc. -- depending on the use case).

You would do this by adding to `write_children(node)` in gen_defines.py using the edtlib.py API, adding the macro itself and doxygen comments to zephyr/devicetree.h, and adding test cases in tests/lib/devicetree/api.

I don't have any plans to implement this but of course I would review patches adding it.

File: drivers/i3c/mux/Kconfig.renesas
Comment (dcpleung) [inline drivers/i3c/mux/Kconfig.renesas:4]: Please use `_INIT_PRIORITY` as suffix so others know exactly what this is for.

File: drivers/i3c/mux/Kconfig.renesas
Comment (dcpleung) [inline drivers/i3c/mux/Kconfig.renesas:28]: Ditto for `_INIT_PRIORITY` suffix.

File: drivers/i3c/mux/Kconfig.renesas
Comment (dcpleung) [inline drivers/i3c/mux/Kconfig.renesas:38]: Possibly a good idea for the above two to depend on this one so the final `.config` would not have the above defined if the mux is not enabled.

**Review by dcpleung — COMMENTED** [review]
Except the kconfigs, LGTM.

File: drivers/i3c/i3c_renesas_imx3112.c
Comment (ramiroderojas) [reply drivers/i3c/i3c_renesas_imx3112.c]: Really sorry for the delay, I was busy with non-zephyr-related matters, and I forgot about the PR.

I removed the SETAASA for now. I think each child driver in the mux can do its own switch to I3C if necessary.
As to the IMX3112 spec sheet/programming guide, I don't think its public yet, although I could ask the support team about that.

I think that being able to expand DT_FOREACH_DESCENDANT(node_id, fn) to do N levels down the tree would be cool, I just don't think I have the Zephyr knowhow to do this :)

**Comment by mfischer** [comment]
Can we get some eyes on this, this is still relevant @cfriedt :)

**Comment by cfriedt** [comment]
> Can we get some eyes on this, this is still relevant @cfriedt :)

I can review, but we are in feature freeze until after the release is cut.

**Comment by ramiroderojas** [comment]
No problem, this can go on next release, thanks @cfriedt :)

**Review by cfriedt — COMMENTED** [review]
Hmm.. I am by no means an i3c expert. Will defer to @XenuIsWatching

**Comment by ramiroderojas** [comment]
I believe I did the changes @XenuIsWatching  requested, I was just waiting on him to take another look

File: drivers/i3c/mux/i3c_renesas_imx3112.c
Comment (XenuIsWatching) [inline drivers/i3c/mux/i3c_renesas_imx3112.c:359]: I don't think thesefunctions are even needed to be defined here at all, as if they are NULL in the api struct, the api will just return -ENOSYS for you anyways

**Comment by XenuIsWatching** [comment]
I only have time to skim through it right now (until I return from paternity leave a couple of weeks from now) rather than going line by line.

Skimming through, I don't see how devices are initialized with a valid DA that does not conflict with anything. How is this done?

**Comment by decsny** [comment]
bot is confused by the "imx", making it think it is an NXP thing, can you please address this somehow in the maintainer yml to indicate it is a renesas device and not NXP

**Pull request closed without merging**
